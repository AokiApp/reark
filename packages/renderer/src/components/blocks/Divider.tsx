import { BlockInnerComponent } from "../../types";

export const Divider: BlockInnerComponent = () => {
  return <hr className="reark-divider" />;
};
