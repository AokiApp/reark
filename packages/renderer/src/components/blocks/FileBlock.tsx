import { BlockInnerComponent } from "../../types";

/**
 * FileBlock: Renders a file block (block_type: 23).
 * - Shows file name and download link if available.
 * - This block is usually paired with a ViewBlock (block_type: 33) for visual representation.
 * - If rendered directly, provides basic file info.
 */
export const FileBlock: BlockInnerComponent = ({ block }) => {
  const file = block.file;
  if (!file) {
    return (
      <div className="reark-file-block reark-file-block--empty">
        No file data
      </div>
    );
  }

  return (
    <div className="reark-file-block">
      <div>
        <strong>File:</strong> {file.name || "(no name)"}
      </div>
      {file.token ? (
        <a
          href={`https://open.feishu.cn/drive/file/${file.token}`}
          target="_blank"
          rel="noopener noreferrer"
          className="reark-file-block__file-link"
        >
          Download / Open
        </a>
      ) : (
        <span className="reark-file-block__file-link--disabled">
          (No file token)
        </span>
      )}
    </div>
  );
};
