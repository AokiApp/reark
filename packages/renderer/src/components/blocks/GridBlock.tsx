import { BlockInnerComponent } from "../../types";
import { BlockComponent } from "../BlockComponent";

/**
 * GridBlock: Renders a grid container with a specified number of columns.
 * Each child is expected to be a GridColumnBlock.
 */
export const GridBlock: BlockInnerComponent = ({ block }) => {
  const grid = block.grid;
  if (!grid || !block.children || !Array.isArray(block.children)) {
    return null;
  }

  // column_size: number of columns (2-4)
  //const columnSize = grid.column_size || 2;

  // Render children by block reference (blockId), allowing for block reference resolution.
  return (
    <div className="reark-grid-wrapper">
      <div
        className="reark-grid"
        style={{
          display: "flex",
          //gridTemplateColumns: `repeat(${columnSize}, 1fr)`,
          gap: "16px",
        }}
      >
        {block.children.map((childId) => (
          // BlockComponent resolves and renders the block by its ID (block reference)
          <BlockComponent key={childId} blockId={childId} />
        ))}
      </div>
    </div>
  );
};
