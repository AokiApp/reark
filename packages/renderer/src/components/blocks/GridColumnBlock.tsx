import { BlockInnerComponent } from "../../types";
import { BlockComponent } from "../BlockComponent";

/**
 * GridColumnBlock: Renders a single column in a grid, with optional width ratio.
 * Children can be any block type.
 */
export const GridColumnBlock: BlockInnerComponent = ({ block }) => {
  const gridColumn = block.grid_column;
  if (!block.children || !Array.isArray(block.children)) {
    return null;
  }

  // width_ratio: ratio of this column to the whole grid (1-99)
  const widthRatio = gridColumn?.width_ratio || 1;

  // Render children by block reference (blockId), allowing for block reference resolution.
  return (
    <div
      className="reark-grid-column"
      style={{
        flex: widthRatio,
        minWidth: 0,
        display: "flex",
        flexDirection: "column",
        gap: "8px",
      }}
    >
      {block.children.map((childId) => (
        // BlockComponent resolves and renders the block by its ID (block reference)
        <BlockComponent key={childId} blockId={childId} />
      ))}
    </div>
  );
};
