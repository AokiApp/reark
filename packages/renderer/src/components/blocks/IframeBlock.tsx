import React from "react";
import { BlockInnerComponent } from "../../types";

function decodeUrl(url: string): string {
  try {
    return decodeURIComponent(url);
  } catch {
    return url;
  }
}

export const IframeBlock: BlockInnerComponent = ({ block }) => {
  const component = block.iframe?.component;
  if (!component) return null;

  const { iframe_type, url } = component;
  const decodedUrl = decodeUrl(url);

  if (iframe_type === 15) {
    // YouTube埋め込み
    // URL例: https://www.youtube.com/watch?v=xxxx
    // 埋め込み用URLに変換
    let embedUrl = "";
    const match = decodedUrl.match(
      /(?:youtube\.com\/watch\?v=|youtu\.be\/)([A-Za-z0-9_-]+)/,
    );
    if (match && match[1]) {
      embedUrl = `https://www.youtube.com/embed/${match[1]}`;
    } else {
      embedUrl = decodedUrl;
    }
    return (
      <div className="reark-iframe">
        <iframe
          src={embedUrl}
          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
          title="YouTube Video"
        />
      </div>
    );
  } else {
    // 汎用iframe
    return (
      <div className="reark-iframe">
        <iframe
          src={decodedUrl}
          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
          title="Embedded Content"
        />
      </div>
    );
  }
};
