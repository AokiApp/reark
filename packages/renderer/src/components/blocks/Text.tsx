import { BlockInnerComponent } from "../../types";
import type { TextElement, TextStyle } from "@aokiapp/reark-lark-api";
import { containsUrl } from "../../utils/utils";
import { Comment } from "../Comment";

export const Text: React.FC<{
  elements: TextElement[];
  style?: TextStyle;
}> = ({ elements, style }) => {
  const align = style?.align || 1;
  const textAlign =
    align === 1
      ? "left"
      : align === 2
        ? "center"
        : align === 3
          ? "right"
          : "left";

  return (
    <div style={{ textAlign }} className="reark-text-block">
      {elements.map((element, index) => {
        if (!element.text_run) {
          return null;
        }

        const textElementStyle = element.text_run.text_element_style;

        // Build className for text decorations
        const classNames = [
          textElementStyle?.bold ? "reark-text--bold" : "",
          textElementStyle?.italic ? "reark-text--italic" : "",
          textElementStyle?.strikethrough ? "reark-text--strikethrough" : "",
          textElementStyle?.underline ? "reark-text--underline" : "",
          textElementStyle?.inline_code ? "reark-text--inline-code" : "",
          "reark-text",
        ]
          .filter(Boolean)
          .join(" ");

        //文字列にリンクが紐づいているor文字列がhttps://で始まる場合はリンクとして扱う
        let url = textElementStyle?.link?.url;
        let isUrl = false;
        if (url) {
          url = decodeURIComponent(url);
          isUrl = true;
        } else if (containsUrl(element.text_run.content)) {
          url = element.text_run.content;
          isUrl = true;
        }

        const inner = (
          <span key={index} className={classNames}>
            {isUrl ? (
              <a href={url} target="_blank">
                {element.text_run.content}
              </a>
            ) : (
              element.text_run.content
            )}
          </span>
        );

        //コメントが紐づけられている場合は、文字列にマーカーを付与
        const comment = textElementStyle?.comment_ids;

        if (Array.isArray(comment) && comment.length > 0) {
          return (
            <Comment commentIds={comment} key={index}>
              {inner}
            </Comment>
          );
        }

        return inner;
      })}
    </div>
  );
};

export const TextBlock: BlockInnerComponent = ({ block }) => {
  return (
    <Text
      elements={block.text?.elements as unknown as TextElement[]} // todo: safer way
      style={block.text?.style}
    />
  );
};
