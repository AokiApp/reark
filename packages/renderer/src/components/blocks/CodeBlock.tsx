import { BlockInnerComponent } from "../../types";
import { CODE_LANGUAGE } from "../../constants/codeLanguage";
import { Text } from "./Text";

export const CodeBlock: BlockInnerComponent = ({ block }) => {
  const code = block.code;
  if (!code) {
    return null;
  }
  const codeLanguage = code?.style?.language
    ? CODE_LANGUAGE[code.style.language]
    : "Unknown";

  return (
    <div className="reark-codeblock">
      <div className="reark-codeblock__language">{codeLanguage}</div>
      <pre className="reark-codeblock__pre">
        <Text elements={code.elements} style={code.style} />
      </pre>
    </div>
  );
};
// todo : copy button
