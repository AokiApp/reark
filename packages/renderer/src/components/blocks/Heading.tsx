import { BlockInnerComponent } from "../../types";
import { Text } from "./Text";

type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

const HeadingBase: BlockInnerComponent<{
  level: HeadingLevel;
}> = ({ block, level }) => {
  const headingKey = `heading${level}` as const;
  const headingData = block[headingKey];

  if (!headingData?.elements) {
    return null;
  }

  return (
    <div
      className={`reark-heading reark-heading--${level}`}
      id={block.block_id}
    >
      <Text elements={headingData.elements} style={headingData.style} />
    </div>
  );
};

export const Heading1: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={1} />
);
export const Heading2: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={2} />
);
export const Heading3: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={3} />
);
export const Heading4: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={4} />
);
export const Heading5: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={5} />
);
export const Heading6: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={6} />
);
export const Heading7: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={7} />
);
export const Heading8: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={8} />
);
export const Heading9: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={9} />
);
