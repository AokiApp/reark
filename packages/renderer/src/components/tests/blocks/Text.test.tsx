import { render, screen } from "@testing-library/react";
import { Text } from "../../blocks/Text";
import { BlockStoreProvider } from "../../../contexts/BlockStoreContext";

// .private.local/example-blocks.json よりTextブロックの例

describe("Text block", () => {
  it("renders text from context", () => {
    const textBlock = {
      block_id: "T8F8dOwtNom7yHxww2ZjPnw3pSf",
      block_type: 2,
      comment_ids: ["7500745405299163169"],
      parent_id: "ZIjadstYfoQVMjxXAwRjM0rVpVg",
      text: {
        elements: [
          {
            text_run: {
              content: "本文書は、 ",
              text_element_style: {
                bold: false,
                comment_ids: ["7500745405299163169"],
                inline_code: false,
                italic: false,
                strikethrough: false,
                underline: false,
              },
            },
          },
          {
            text_run: {
              content: "lark2react",
              text_element_style: {
                bold: false,
                comment_ids: ["7500745405299163169"],
                inline_code: true,
                italic: false,
                strikethrough: false,
                underline: false,
              },
            },
          },
          {
            text_run: {
              content: "の描画を確かめるためのものである。",
              text_element_style: {
                bold: false,
                comment_ids: ["7500745405299163169"],
                inline_code: false,
                italic: false,
                strikethrough: false,
                underline: false,
              },
            },
          },
        ],
        style: {
          align: 1 as const,
          folded: false,
        },
      },
    };
    const { container } = render(
      <BlockStoreProvider items={[textBlock]}>
        <Text elements={textBlock.text.elements} style={textBlock.text.style} />
      </BlockStoreProvider>,
    );
    // 主要なテキストが表示されていることを確認（部分一致・正規表現でマッチ）
    expect(screen.getByText(/本文書は/)).toBeInTheDocument();
    expect(screen.getByText(/lark2react/)).toBeInTheDocument();
    expect(
      screen.getByText(/の描画を確かめるためのものである/),
    ).toBeInTheDocument();
    // スナップショットテスト
    expect(container).toMatchSnapshot();
  });
});
