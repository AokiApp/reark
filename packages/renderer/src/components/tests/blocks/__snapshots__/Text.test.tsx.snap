// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text block renders text from context 1`] = `
<div>
  <div
    class="reark-text-block"
    style="text-align: left;"
  >
    <div
      class="reark-comment"
      data-comment-ids="7500745405299163169"
    >
      <span
        class="reark-text"
      >
        本文書は、 
      </span>
    </div>
    <div
      class="reark-comment"
      data-comment-ids="7500745405299163169"
    >
      <span
        class="reark-text--inline-code reark-text"
      >
        lark2react
      </span>
    </div>
    <div
      class="reark-comment"
      data-comment-ids="7500745405299163169"
    >
      <span
        class="reark-text"
      >
        の描画を確かめるためのものである。
      </span>
    </div>
  </div>
</div>
`;

exports[`renders empty span when elements contain empty string 1`] = `
<div>
  <div
    class="reark-text-block"
    style="text-align: left;"
  >
    <span
      class="reark-text"
    />
  </div>
</div>
`;

exports[`renders text with bold style 1`] = `
<div>
  <div
    class="reark-text-block"
    style="text-align: left;"
  >
    <span
      class="reark-text--bold reark-text"
    >
      太字
    </span>
    <span
      class="reark-text"
    >
      は英語で
    </span>
    <span
      class="reark-text--bold reark-text"
    >
      Bold
    </span>
  </div>
</div>
`;
